energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+
+    m_activePowerStateTypeIds.insert(pro380ThingClassId, pro380CurrentPowerEventTypeId);
+    m_activePowerStateTypeIds.insert(sdm630ThingClassId, sdm630CurrentPowerStateTypeId);
+
+    m_frequencyStateTypeIds.insert(pro380ThingClassId, pro380FrequencyStateTypeId);
+    m_frequencyStateTypeIds.insert(sdm630ThingClassId, sdm630FrequencyStateTypeId);
+
+    m_powerFactorStateTypeIds.insert(pro380ThingClassId, pro380PowerFactorStateTypeId);
+    m_powerFactorStateTypeIds.insert(sdm630ThingClassId, sdm630PowerFactorStateTypeId);
+
+    m_totalEnergyConsumedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyConsumedEventTypeId);
+    m_totalEnergyConsumedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyConsumedEventTypeId);
+
+    m_totalEnergyProducedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyProducedStateTypeId);
+    m_totalEnergyProducedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyProducedStateTypeId);
+
+    m_discoverySlaveAddressParamTypeIds.insert(pro380ThingClassId, pro380DiscoverySlaveAddressParamTypeId);
+    m_discoverySlaveAddressParamTypeIds.insert(sdm630ThingClassId, sdm630DiscoverySlaveAddressParamTypeId);
+
+    m_registerMaps.insert(pro380ThingClassId, pro380RegisterMap);
+    m_registerMaps.insert(sdm630ThingClassId, sdm630RegisterMap);
+
+    // FIXME leads to crash, probably an issue in the modbus rtu resource
+    /*connect(hardwareManager()->modbusRtuResource(), &ModbusRtuHardwareResource::modbusRtuMasterRemoved, this, [=] (const QUuid &modbusUuid){
+        qCDebug(dcEnergyMeters()) << "Modbus RTU master has been removed" << modbusUuid.toString();
+
+        // Check if there is any device using this resource
+        foreach (EnergyMeter * meter, m_energyMeters) {
+            if (meter->modbusRtuMasterUuid() == modbusUuid) {
+                Thing *thing = m_energyMeters.key(meter);
+                if (!thing)
+                    return;
+
+                qCWarning(dcEnergyMeters()) << "Modbus RTU hardware resource removed for" << thing << ". The thing will not be functional any more until a new resource has been configured for it.";
+                thing->setStateValue(m_connectionStateTypeIds[thing->thingClassId()], false);
+            }
+        }
+    }); */
+}
+
+void IntegrationPluginEnergyMeters::discoverThings(ThingDiscoveryInfo *info)
+{
+    qCDebug(dcEnergyMeters()) << "Discover things";
+    QList<ThingDescriptor> thingDescriptors;
+
+    if (hardwareManager()->modbusRtuResource()->modbusRtuMasters().isEmpty()) {
+        info->finish(Thing::ThingErrorHardwareNotAvailable, QT_TR_NOOP("No Modbus RTU interface available."));
+        return;
+    }
+
+    if (m_connectionStateTypeIds.contains(info->thingClassId())) {
+        int slaveAddress = info->params().paramValue(m_discoverySlaveAddressParamTypeIds.value(info->thingClassId())).toInt();
+        if (slaveAddress > 254 || slaveAddress == 0) {
```

Comment:
What about negativ slave addresses? Either you define the slave address as `UInt` type in the plugin json, or handle it here.